KIPs/kip-draft_kgp6.md

```diff
@@ -0,0 +1,652 @@
+        } else {
+            //check if the msg.sender is one of the admin of the retiredAddress contract
+            require(
+                _validateAdmin(_retiredAddress, msg.sender),
+                "msg.sender is not the admin"
+            );
+            _updateApprover(_retiredAddress, msg.sender);
+        }
+    }
+
+    /**
+     * @dev validate if the msg.sender is admin if the retiredAddress is a contract
+     * @param _retiredAddress is the address of the contract
+     * @param _approver is the msg.sender
+     * @return isAdmin is true if the msg.sender is one of the admin
+     */
+    function _validateAdmin(
+        address _retiredAddress,
+        address _approver
+    ) private returns (bool isAdmin) {
+        (address[] memory adminList, ) = _getState(_retiredAddress);
+        require(adminList.length != 0, "admin list cannot be empty");
+        for (uint8 i = 0; i < adminList.length; i++) {
+            if (_approver == adminList[i]) {
+                isAdmin = true;
+            }
+        }
+    }
+
+    /**
+     * @dev gets the adminList and quorom by calling `getState()` method in retiredAddress contract
+     * @param _retiredAddress is the address of the contract
+     * @return adminList list of the retiredAddress contract admins
+     * @return req min required number of approvals
+     */
+    function _getState(
+        address _retiredAddress
+    ) private returns (address[] memory adminList, uint256 req) {
+        //call getState() function in retiredAddress contract to get the adminList
+        bytes memory payload = abi.encodeWithSignature("getState()");
+        (bool success, bytes memory result) = _retiredAddress.staticcall(
+            payload
+        );
+        emit GetState(success, result);
+        require(success, "call failed");
+
+        (adminList, req) = abi.decode(result, (address[], uint256));
+    }
+
+    /**
+     * @dev Internal function to update the approver details of a retired
+     * _retiredAddress is the address of the retired
+     * _approver is the admin of the retiredAddress
+     */
+    function _updateApprover(
+        address _retiredAddress,
+        address _approver
+    ) private {
+        uint256 index = getRetiredIndex(_retiredAddress);
+        address[] memory approvers = retirees[index].approvers;
+        for (uint256 i = 0; i < approvers.length; i++) {
+            require(
+                approvers[i] != _approver,
+                "Already approved"
+            );
+        }
+        retirees[index].approvers.push(_approver);
+        emit Approved(
+            _retiredAddress,
+            _approver,
+            retirees[index].approvers.length
+        );
+    }
+
+    /**
+     * @dev finalizeRegistration sets the status to Registered,
+     *      After this stage, registrations will be restricted.
+     */
+    function finalizeRegistration()
+        public
+        onlyOwner
+        onlyAtStatus(Status.Initialized)
+    {
```

Comment:
Who calls this function? by core or by owner?
If it is called by owner, don't we need to validate whether we can finalize it or not?

Reply:
It is called by owner.  How can we finalize whether registration is complete ? @aidan-kwon